### PR TITLE
Add extensions priority in sonata_admin config

### PR DESCRIPTION
**Pull request by EmmanuelVella**
I am targeting this branch, because it's a BC new feature.

## Changelog

```markdown
### Added
- Added a `priority` option in `sonata_admin` extensions config
```

## To do

- [x] Update the tests
- [x] Update the documentation

## Subject

This PR adds a `priority` option in the `sonata_admin` extensions config.

This [previous PR](https://github.com/sonata-project/SonataAdminBundle/pull/4298) introduced the `priority` option in services tags, but the `SplPriorityQueue` class used **do not preserve the insertion order** in case of same priority (so the extensions declaration order is not preserved). I'm not sure if it should be considered as a bug, but the new `priority` option now allows to finely configure the order.

Note : I fixed the "`get_class($this->extensionMock)`" calls in the tests : we can't rely on it since it returns the same mock class for all the extensions mocks.

### CODE REVIEW

File: Resources/doc/reference/extensions.rst [hunk 1]
```diff
@@ -100,6 +100,9 @@ uses:
     Requires PHP >= 5.4.0. Specify one or more traits. If the managed class of an admin uses one of the specified traits the extension will be
     added to that admin.
 
+priority:
+    Can be a positive or negative integer. The higher the priority, the earlier it’s executed.
```
Comment (OskarStark) [inline Resources/doc/reference/extensions.rst]: Is this the same order as the priority is used in Symfony itself?

**Comment by OskarStark** [comment]
I like this PR 👍

File: DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 2]
```diff
@@ -181,12 +181,14 @@ protected function flattenExtensionConfiguration(array $config)
...
-            foreach ($options as $key => $value) {
+            $optionsMap = array_intersect_key($options, $extensionMap);
+
+            foreach ($optionsMap as $key => $value) {
                 foreach ($value as $source) {
                     if (!isset($extensionMap[$key][$source])) {
                         $extensionMap[$key][$source] = array();
                     }
-                    array_push($extensionMap[$key][$source], $extension);
+                    $extensionMap[$key][$source][$extension]['priority'] = $options['priority'];
```
Comment (greg0ire) [inline DependencyInjection/Compiler/ExtensionCompilerPass.php]: Please change the phpdoc for the return value of this method, which is now completely different unless I'm mistaken

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -425,6 +425,9 @@ public function getConfigTreeBuilder()
                                     ->thenInvalid('PHP >= 5.4.0 is required to use traits.')
                                 ->end()
                             ->end()
+                            ->integerNode('priority')
+                                ->defaultValue(0)
```
Comment (greg0ire) [inline DependencyInjection/Configuration.php]: Please document this with a call to `info()`, for instance does 0 mean high or low priority?

File: Resources/doc/reference/extensions.rst (hunk 1 above)
Comment (EmmanuelVella) [inline Resources/doc/reference/extensions.rst]: Yes ! For example, a 128 priority will be exectued before a 0 or a -255 priority.

File: DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 2 above)
Comment (EmmanuelVella) [inline DependencyInjection/Compiler/ExtensionCompilerPass.php]: It still returns an array. Would you like me to detail it's structure ?

Comment (greg0ire) [inline DependencyInjection/Compiler/ExtensionCompilerPass.php]: Yes. What are the keys? What are the values? How deep is it?

**Comment by EmmanuelVella** [comment]
PR updated and tests are ok now !

File: DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 4]
```diff
@@ -165,6 +165,17 @@ protected function getManagedClass(Definition $admin, ContainerBuilder $containe
...
+     *     'admins'     => array('<admin_id>'  => array('<extension_id>' => array('priority' => <int>))),
+     *     'implements' => array('<interface>' => array('<extension_id>' => array('priority' => <int>))),
+     *     'extends'    => array('<class>'     => array('<extension_id>' => array('priority' => <int>))),
+     *     'instanceof' => array('<class>'     => array('<extension_id>' => array('priority' => <int>))),
+     *     'uses'       => array('<trait>'     => array('<extension_id>' => array('priority' => <int>))),
+     * )
+     *
      * @param array $config
      *
      * @return array
```
Comment (greg0ire) [inline DependencyInjection/Compiler/ExtensionCompilerPass.php]: Way more clear! Can you move the description here?

**Comment by OskarStark** [comment]
@EmmanuelVella can you please check travis?

**Comment by greg0ire** [comment]
@OskarStark travis is blocked because of https://github.com/sonata-project/dev-kit/pull/267

**Comment by EmmanuelVella** [comment]
Tests are ok now ! 👍

**Comment by OskarStark** [comment]
@greg0ire please merge if you are fine with it, thank you!

**Comment by OskarStark** [comment]
Thank you @EmmanuelVella 👍

**Comment by EmmanuelVella** [comment]
Great, thank you !

**Pull request merged**